lib/OpenQA/Scheduler/Model/Jobs.pm

```diff
@@ -132,79 +132,121 @@ sub schedule {
 
     my @successfully_allocated;
 
+    # assign the allocated job-worker pairs
     for my $allocated (values %$allocated_jobs) {
-        #  Now we need to set the worker in the job, with the state in SCHEDULED.
-        my $job;
+        # find worker
+        my $worker_id = $allocated->{worker};
         my $worker;
         try {
-            $job = $schema->resultset("Jobs")->find({id => $allocated->{job}});
+            $worker = $schema->resultset('Workers')->find({id => $worker_id});
         }
         catch {
-            log_debug("Failed to retrieve Job(" . $allocated->{job} . ") in the DB :( bummer! Reason: $_");
+            log_debug("Failed to retrieve worker ($worker_id) in the DB, reason: $_");
         };
+        next unless $worker;
+        if ($worker->unfinished_jobs->count) {
+            log_debug "Worker already got jobs, skipping";
+            next;
+        }
 
+        # take directly chained jobs into account
+        # note: That these jobs have a matching WORKER_CLASS is enforced on dependency creation.
+        my $first_job_id = $allocated->{job};
+        my ($directly_chained_job_sequence, $job_ids)
+          = _serialize_directly_chained_job_sequence($first_job_id, $scheduled_jobs->{$first_job_id}->{cluster_jobs});
+
+        # find jobs
+        my @jobs;
+        my $job_ids_str = join(', ', @$job_ids);
         try {
-            $worker = $schema->resultset("Workers")->find({id => $allocated->{worker}});
+            @jobs = $schema->resultset('Jobs')->search({id => {-in => $job_ids}});
         }
         catch {
-            log_debug("Failed to retrieve Worker(" . $allocated->{worker} . ") in the DB :( bummer! Reason: $_");
+            log_debug("Failed to retrieve jobs ($job_ids_str) in the DB, reason: $_");
         };
-
-        next unless $job && $worker;
-        if ($worker->job) {
-            log_debug "Worker already got a job, skipping";
+        my $actual_job_count = scalar @jobs;
+        if ($actual_job_count != scalar @$job_ids) {
+            log_debug("Failed to retrieve jobs ($job_ids_str) in the DB, reason: only got $actual_job_count jobs");
             next;
         }
-        if ($job->state ne SCHEDULED) {
-            log_debug "Job no longer scheduled, skipping";
-            next;
+
+        # check whether the jobs are still scheduled
+        my $skip = 0;
+        for my $job (@jobs) {
+            my $job_id = $job->id;
+            if ($job->state ne SCHEDULED) {
+                log_debug("Job $job_id no longer scheduled, skipping");
+                $skip = 1;
+                last;
+            }
         }
+        next if $skip;
```

Comment:
how about something like `next if grep { $_->state ne SCHEDULED} @jobs;`

If you still want to have the log message:

```perl
if (my @skipped_jobs = grep { $_->state ne SCHEDULED} @jobs) {
    log_debug("Job $_->id no longer scheduled, skipping") foreach @skipped_jobs;
    next;
}
```